src/clr-angular/utils/smart-popover/providers/smart-popover-position.service.ts

```diff
@@ -0,0 +1,196 @@
+/*
+ *  Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ *
+ */
+import { isPlatformBrowser } from '@angular/common';
+import { Injectable, PLATFORM_ID, Inject } from '@angular/core';
+
+import { ClrSmartPopoverEventsService } from './smart-popover-events.service';
+import { ClrAlignment } from '../enums/alignment.enum';
+import { ClrSmartPosition } from '../interfaces/smart-position.interface';
+import { ClrPopoverContentOffset } from '../interfaces/popover-content-offset.interface';
+import { ClrViewportViolation } from '../enums/viewport-violation.enum';
+import { align, flipSidesAndNudgeContent, flipSides, nudgeContent, testVisibility } from '../clr-position-operators';
+import { ClrAxis } from '../enums/axis.enum';
+
+@Injectable()
+export class ClrSmartPopoverPositionService {
+  private currentAnchorCoords: ClientRect;
+  private currentContentCoords: ClientRect;
+  private contentOffsets: ClrPopoverContentOffset;
+  private _position: ClrSmartPosition;
+
+  set position(position: ClrSmartPosition) {
+    this._position = position;
+  }
+  get position(): ClrSmartPosition {
+    return this._position;
+  }
+
+  constructor(private eventService: ClrSmartPopoverEventsService, @Inject(PLATFORM_ID) public platformId: Object) {}
+
+  public alignContent(content: HTMLElement): ClrPopoverContentOffset {
+    if (!isPlatformBrowser(this.platformId)) {
+      // Only position when in a browser.
+      // Default to the browser origin and prevent getBoundingClientRect from running.
+      return {
+        xOffset: 0,
+        yOffset: 0,
+      };
+    }
+
+    this.currentAnchorCoords = this.eventService.anchorButtonRef.nativeElement.getBoundingClientRect();
+    this.currentContentCoords = content.getBoundingClientRect();
+    this.contentOffsets = align(this.position, this.currentAnchorCoords, this.currentContentCoords);
+
+    const visibilityErrors: ClrViewportViolation[] = testVisibility(this.contentOffsets, this.currentContentCoords);
+    const errorCode = visibilityErrors.reduce((count, current) => {
+      return count + current;
+    }, 0);
+
+    // DOCUMENT What cases are described here. Is there a way to make it more descriptive
+    // seems clever needs explained.
```

Comment:
Agreed, this is too clever and not obvious without more detail. Could you also map error codes to an enum?

Reply:
Addressed this with some inline comments and breaking out the switches into descriptive method names that tell what the switch does for each if branch.